Make use of Xilinx-specific clock features and constraints

andrewsil1 opened the pull request:
These changes significantly clean up and fully constrain all paths in the Nexys4 project, and eliminate the use of the tim.v and P1V.v modules in favor of an Artix-7 specific implementation that makes full use of the MMCM clocking module for all modes except RCSLOW and divisor-based PLL simulation.

The Nexys4 constraints also include PBlock floorplanning that aids somewhat in speeding up the implementation process, as well as tying Pin10 to the Nexys4 audio amplifier PWM input and the amplifier enable to Pin10's direction such that audio is enabled when Pin10 is an output.  This allows the Propeller "Singing" demos to work properly.

andrewsil1 committed:
Add asynchronous synchronizers for all Propeller input pins

All of the Propeller inputs are effectively asynchronous to the clocks driving the cogs.
In order to minimize the possibility of metastability, standard practice for asynchronous
data inputs to a flop is to pass them through a minimum of two, and preferably more, layers of
chained flops, clocked in the domain of the destination logic, at the expense of a few cycles of input latency.

andrewsil1 committed:
Merge branch 'jacgoudsmit/xilinx' into xilinx

andrewsil1 committed:
Latest in-progress attempts at async input synchronizers.

andrewsil1 committed:
Merge branch 'rel' into xilinx

andrewsil1 committed:
Local change

andrewsil1 committed:
Revise clocking strategy to make full use of Xilinx MMCM

Rather than use logic-based counters to generate divided clocks, do it with the Xilinx-specific blocks.
This aids in timing and ensures a strict phase relationship between all of the clocks.

Also modifies the video section to synthesize more cleanly by using the enable signal as a clock enable
on the vclk, rather than driving the clock net with a LUT that combines them with logic.

andrewsil1 committed:
General code cleanup, completion of MMCM clock generation changes, and added pblocks to constraints to help with Nexys4 placement.
Also changes tree structure to eliminate "include" statements, make all .V files first class citizens in the project.

andrewsil1 committed:
By registering the output from the "fake PLL", DRC errors are eliminated where it doesn't like the VCLK in cog_vid being sourced by the output of a combinational pin.  Sticking a register in between the combinational logic and the usage of VCLK eliminates the warning and also significantly speeds up the routing phase of implementation, at the expense of one-half-cogclock of latency in the PLL counter, which should be irrelevant.

andrewsil1 committed:
Remove usage of the P1V.v and tim.v modules in their entirety, in the following fashion:
- Instantiate dig.v directly from the Nexys4 top module
- Move all clock generation to the Xilinx-specific clock generator, eliminating the use of divider-generated clocks in all cases except RCSLOW (20khz.)
- Eliminate the unnecessary LED output generate section in the top, just map the 8-bit vector directly from dig.v to the IO port.

andrewsil1 committed:
Revert private changes to tim.v, since it's no longer needed for the Xilinx Nexys4 implementation.

andrewsil1 committed:
Revert private changes to p1v.v since it's no longer needed for the Nexys4.

andrewsil1 committed:
Merge rel changes to unused modules into xilinx branch.

andrewsil1 committed:
Bring core modules up to date with rel (removal of include statements)

andrewsil1 committed:
Significant cleanup and enhancement of Nexys constraints

All clock paths are now fully described and constrained, including all
generated clocks such as RCSLOW, PLLA and PLLB for each cog, and VCLK
(video clock derived from PLLA in each cog.)

Added comments and cleaned up the constraints file to make it clearer
what all this stuff is doing.

Also now generating a 128Mhz clock in the MMCM directly, to act as the
pseudo-PLL rather than the previous 160Mhz version.

andrewsil1 committed:
Merge branch 'xilinx' into xilinx

jacgoudsmit commented:
Hey Andy,

I merged the changes to delete the include commands and to add files to the projects. 

But I have a few concerns about the changes that you proposed after that: 

- I noticed you separated "vclk" from "enable" in the cog_vid.v module, and I see how that would work in the light of earlier changes from you, so I have no problem with those. But you also removed "or negedge ena" from an `@always` clause around line 52, and you removed an "or posedge snc[1]"  from an `@always` clause around line 107. I don't claim to be an expert and I may don't completely grok the cog_vid module, but I would like to know why those changes are necessary.
- There were some minor layout issues in various places, for example it looked like you accidentally pasted an entire commented-out clock generator in the middle of a comment, and there were some tab characters in the source.
- You added an entity "simtop" 
- I do **NOT** want a binary file in my tree that is normally generated from the source code. Binary files use up unnecessary disk space and bandwidth. Sure, one 1MB binary file isn't much, but it sets a precedent and if we would start added binaries for every target, we'd soon get a Git repo that takes forever to clone or push, because they usually can't be compressed when they get sent over a network. This is what happened in the old P8X32_Emulation repo where someone once made a mistake to check in a ton of binaries.

I carefully merged your recent changes into a new branch "pr11" (for "pull request 11") one by one, except for the one that added the bin file. I undid the two changes to cog_vid and made some minor changes to clean up other code (mostly layout).  Then I merged the changes to your code to check if there was anything obvious that I missed,

Currently at the head of the pr11 branch, the Arty build is broken because you changed the xilinx_clock module and changed the Nexys4 project so it uses dig.v instead of p1v, and I haven't implemented those changes in the Arty project yet. I may have time for that later today or possibly tomorrow.

===Jac

andrewsil1 commented:
I’m perfectly fine with removing the binary checkin if you feel strongly about it.  I’ll also remove the sim module, I had added it to try to debug something but no longer need it and forgot it was in the project.  Also not sure how the tabs worked their way in, I’ve changed my editors to use spaces, I’ll figure out where that came from and try to avoid it in the future.

The “or” clauses are removed for the same reason as previous async reset removals elsewhere in the code below the top level, they’re just turned into synchronous resets by removal from the sensitivity list.  That much is the same as the earlier checkins. The removal of the enable that feeds VCLK helps implementation by not feeding a clock buffer from a combinational component, which adds skew, but instead uses the enable signal as a clock enable at each flop that makes use of VCLK.  Logically identical but better suited to FPGA architectures and the tools whine less and run faster.

I’ll pull down your new branch and clean up anything I messed up. Thanks for the review of the pull.

Sent from Outlook<https://aka.ms/qtex0l> for iOS- Tiny keys, big fingers.
_____________________________
From: Jac Goudsmit <notifications@github.com<mailto:notifications@github.com>>
Sent: Sunday, September 24, 2017 1:38 PM
Subject: Re: [jacgoudsmit/P1V] Make use of Xilinx-specific clock features and constraints (#11)
To: jacgoudsmit/P1V <p1v@noreply.github.com<mailto:p1v@noreply.github.com>>
Cc: Andrew Silverman <andrewsi@microsoft.com<mailto:andrewsi@microsoft.com>>, Author <author@noreply.github.com<mailto:author@noreply.github.com>>



Hey Andy,

I merged the changes to delete the include commands and to add files to the projects.

But I have a few concerns about the changes that you proposed after that:

  *   I noticed you separated "vclk" from "enable" in the cog_vid.v module, and I see how that would work in the light of earlier changes from you, so I have no problem with those. But you also removed "or negedge ena" from an@always clause around line 52, and you removed an "or posedge snc[1]" from an@always clause around line 107. I don't claim to be an expert and I may don't completely grok the cog_vid module, but I would like to know why those changes are necessary.
  *   There were some minor layout issues in various places, for example it looked like you accidentally pasted an entire commented-out clock generator in the middle of a comment, and there were some tab characters in the source.
  *   You added an entity "simtop"
  *   I do NOT want a binary file in my tree that is normally generated from the source code. Binary files use up unnecessary disk space and bandwidth. Sure, one 1MB binary file isn't much, but it sets a precedent and if we would start added binaries for every target, we'd soon get a Git repo that takes forever to clone or push, because they usually can't be compressed when they get sent over a network. This is what happened in the old P8X32_Emulation repo where someone once made a mistake to check in a ton of binaries.

I carefully merged your recent changes into a new branch "pr11" (for "pull request 11") one by one, except for the one that added the bin file. I undid the two changes to cog_vid and made some minor changes to clean up other code (mostly layout). Then I merged the changes to your code to check if there was anything obvious that I missed,

Currently at the head of the pr11 branch, the Arty build is broken because you changed the xilinx_clock module and changed the Nexys4 project so it uses dig.v instead of p1v, and I haven't implemented those changes in the Arty project yet. I may have time for that later today or possibly tomorrow.

===Jac

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjacgoudsmit%2FP1V%2Fpull%2F11%23issuecomment-331738642&data=02%7C01%7Candrewsi%40microsoft.com%7Cd412dbdcf2454e12e65d08d5038c3ebc%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636418823237812544&sdata=oTEt46oqkTHGROIq%2F2nkfXJhxSTjrY5yZntf69VI%2BKA%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAGFTLRjavdwvADlZrlVPjXJx8gO7edFDks5slr3RgaJpZM4PhoOF&data=02%7C01%7Candrewsi%40microsoft.com%7Cd412dbdcf2454e12e65d08d5038c3ebc%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636418823237812544&sdata=bC0R8biPehCMvE13lwiFzOA7TZCrO2MRh9lR3NB8tZY%3D&reserved=0>.

jacgoudsmit commented:
Thanks!

Keep in mind that it's not enough to just do a delete-checkin of the bin file: you have to hard-reset the branch to a point just before you added the bin file, and then force-push the branch to github. That will get rid of the changes after you added the bin file too, so you may want to make a local copy (not a stash) of those changes so you can re-apply them afterwards. There may be other ways to do things like that with Git but that's the way I understand it.

andrewsil1 committed:
Merge branch 'rel' of https://github.com/jacgoudsmit/P1V into xilinx

andrewsil1 commented:
FYI, I have forced my fork (and therefore this pull request) back to being up-to-date minus the bitfile checkin. I think it should be in a good state.
